webpack/merge-deep.js

```diff
@@ -0,0 +1,41 @@
+function isObject(obj) {
+  return obj !== null && typeof obj === 'object';
+}
+
+function isPlainObject(obj) {
+  return isObject(obj) && (
+          obj.constructor === Object  // obj = {}
+          || obj.constructor === undefined // obj = Object.create(null)
+        );
+}
+
+function mergeDeep(target, ...sources) {
+  if (!sources.length) { return target; }
+  const source = sources.shift();
+
+  if(Array.isArray(target)) {
+    if(Array.isArray(source)) {
+      target.push(...source);
+    } else {
+      target.push(source);
+    }
+  } else if(isPlainObject(target)) {
+    if(isPlainObject(source)) {
+      for(let key of Object.keys(source)) {
+        if(!target[key]) {
+          target[key] = source[key];
+        } else {
+          mergeDeep(target[key], source[key]);
+        }
+      }
+    } else {
+      throw new Error(`Cannot merge object with non-object`);
+    }
+  } else {
+    target = source;
+  }
+
+  return mergeDeep(target, ...sources);
+};
+
+module.exports = mergeDeep;
```

Comment:
I would just use Lodash here, especially since it's only used on the "server" or locally.

Reply:
In general you don't need to deep merge these webpack config objects, it's often enough to just use a shallow copy and replace specific bits. It's a mess trying to keep these clean, though.

Reply:
Yeah I might do just that. This `mergeDeep` thing just fixed the annoying problem I'd hit so I left it for now. Will replace with lodash